microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/properties/PropertiesFileTextDocumentService.java

```diff
@@ -456,4 +479,39 @@ public <R> CompletableFuture<R> getPropertiesModelCompose(TextDocumentIdentifier
 		return documents.computeModelAsyncCompose(documentIdentifier, code);
 	}
 
+	/**
+	 * Returns the minimal data to resolve a completion item.
+	 *
+	 * @param propertyName the property name of the completion item.
+	 * @param uri          the uri of the text document.
+	 * @return the minimal data to resolve a completion item.
+	 */
+	public static JsonObject createCompletionData(String propertyName, String uri) {
+		JsonObject data = new JsonObject();
+		data.addProperty(DATA_PROPERTY_NAME_FIELD, propertyName);
+		data.addProperty(DATA_URI_FIELD, uri);
+		return data;
+	}
+
+	/**
+	 * Returns the property value of the given <code>fieldName</code> if the given
+	 * <code>data</code> and null otherwise.
+	 *
+	 * @param data      the data entry field comings from the CompletionItem,
+	 *                  CodeAction, etc.
+	 *
+	 * @param fieldName the entry field name (ex : "uri", "propertyName").
+	 *
+	 * @return the property value of the given <code>fieldName</code> if the given
+	 *         <code>data</code> and null otherwise.
+	 */
+	public static String getProperty(Object data, String fieldName) {
+		JsonObject json = JSONUtility.toModel(data, JsonObject.class);
```

Comment:
Instead of doing this, you can make a Java class called `CompletionData` with fields `propertyName` and `uri` and a no-argument constructor.

Then you can call `JSONUtility.toModel(data, CompletionData.class)`, and it will return an instance of `CompletionData`.

There is an example of this at the end of `InitializationOptionsSettings.java`.

I think this will make it a bit easier to read the code.

Reply:
I agree with @datho7561 it is better to do that since completion support is not extensible with participant like lemminx.